### ISSUE TITLE
Support adding a datacenter with tablets by changing rf from 0->N

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#17572]
Currently, adding a datacenter is done in 3 phases:
1. Adding nodes to the new datacenter. At this point the replication factor for that dc is 0 in the network-topology strategy options.
2. ALTER KEYSPACE to set the replication factor in the new dc to some RF>0.
3. Running `nodetool rebuild <existing dc>` on all nodes in the new dc to rebuild the newly added replicas in place.

Depends on:

- [ ] https://github.com/scylladb/scylladb/issues/16128
- [ ] https://github.com/scylladb/scylladb/issues/16129

**Comment by bhalevy** [comment]
compaction_additional_test.py::TestTimeWindowDataSegregation::test_streaming_on_rebuild_multidc depends on this

**Comment by tzach** [comment]
@bhalevy note that the [current add DC procedure](https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-dc-to-existing-dc.html#add-new-dc)
include two additional steps you did not mentioned:

- updating the RF of system tables: system_auth, system_distributed, system_traces
   Do #16970 make this step redundant ?
- run a full cluster repair after the rebuild

**Comment by bhalevy** [comment]
> @bhalevy note that the [current add DC procedure](https://opensource.docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-dc-to-existing-dc.html#add-new-dc) include two additional steps you did not mentioned:

Yes, you're right. I was referring only to the tablets-related steps.

> 
> * updating the RF of system tables: system_auth, system_distributed, system_traces
>   Do [auth: move system_auth under RAFT (epic) #16970](https://github.com/scylladb/scylladb/issues/16970) make this step redundant ?

Once we move the system-distributed tables to raft, they'll get automatically propagated to new nodes via group0.  Their replication strategy will be local anyhow so increasing the replication factor would not be applicable.

> * run a full cluster repair after the rebuild

Yes, that's still required for vnodes-based keyspaces.

**Comment by xemul** [comment]
I can confirm that with #16723 adding new DC starts to work with tablets. Kudos @ptrsmrn !!!

In particular, the rebuild_test::TestRebuild::test_simple_rebuild passes.
Also, since it's not currently possible to increase RF by more than 1, to add a new DC with large RF one can (and currently should) increase RF for the new DC several times one-by-one. Added a dtest for that, see scylladb/scylla-dtest#4199

This way of changing RF has a nice side effect. If increasing RF by 3 instantly, tablets migration code would stream each tablet from source DC several times -- once for each target node. We wanted to have "hierarchical" cross-DC streaming, but currently it's not there. Increasing RF by 1 several time sort of solves this. First increase would stream tablets into destination DC from source DC once, next increases would stream tablets within the destination DC only

**Comment by bhalevy** [comment]
> @bhalevy - should this also move to 6.1?

Yes.  This requires https://github.com/scylladb/scylladb/issues/16129 that was pushed out to 6.1

**Comment by ptrsmrn** [comment]
> > @bhalevy - should this also move to 6.1?
> 
> Yes. This requires #16129 that was pushed out to 6.1

@mykaul @bhalevy is this decision to be reconsidered, since https://github.com/scylladb/scylladb/pull/16723 has been merged to master OSS and will be backported to 6.0?

**Comment by mykaul** [comment]
@ptrsmrn - is there a PR here?

**Comment by xemul** [comment]
No, adding new DC doesn't need extra patching at this point
Here's a dtest that shows how it works https://github.com/scylladb/scylla-dtest/pull/4199 (it needs a bit more care before merging)

**Comment by bhalevy** [comment]
@xemul I'd like to add support for changing RF from 0 to N in one step (and automate the replica building under the covers).
Would you like to handle this under this issue or open a new one for it?

**Comment by xemul** [comment]
I can handle it here

**Comment by mykaul** [comment]
@bhalevy - what's missing here to complete this?

**Comment by mykaul** [comment]
@tgrabiec - it's unclear what tasks are left here to complete this.

**Comment by dani-tweig** [comment]
@denesb , @Deexie , any update here?

**Comment by Deexie** [comment]
> [@denesb](https://github.com/denesb) , [@Deexie](https://github.com/Deexie) , any update here?

We have https://github.com/scylladb/scylladb/pull/24421, but it didn't update it to the latest version. I do changes to it locally. The change depends on [global queue](https://github.com/gleb-cloudius/scylla/tree/global-queue), [colocated tablets](https://github.com/scylladb/scylladb/pull/22906), and @tgrabiec's rf-rack-list. It seems like the first two will make it to master soon and that's probably when I will push my changes to https://github.com/scylladb/scylladb/pull/24421 as it will be easier to keep them up to date.

**Comment by dani-tweig** [comment]
> It seems like the first two will make it to master soon

mind that 2025.3 will be branched on July 1st.

**Comment by denesb** [comment]
Replaced by SCYLLADB-103.

**Comment by denesb** [comment]
Reopening because this caused another dtest bad test resurrection fiasko.
Caused release / dtest with tablets / compaction_additional_test.TestTimeWindowDataSegregation.test_streaming_on_rebuild_multidc (from dtest-with-tablets) to fail.
E.g. as in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22771/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/release___dtest_with_tablets___test_streaming_on_rebuild_multidc/

**Comment by ptrsmrn** [comment]
@dani-tweig  that reminds me we have to be careful about closing GH issues, as some (d)tests execute differently depending on whether the issue is opened/closed

**Comment by pehala** [comment]
> Reopening because this caused another dtest bad test resurrection fiasko. Caused release / dtest with tablets / compaction_additional_test.TestTimeWindowDataSegregation.test_streaming_on_rebuild_multidc (from dtest-with-tablets) to fail. E.g. as in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22771/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/release___dtest_with_tablets___test_streaming_on_rebuild_multidc/

Correct way is for sure to add a dtest/skip-labels here, which can be done in advance even before merging the PR

**Comment by denesb** [comment]
> > Reopening because this caused another dtest bad test resurrection fiasko. Caused release / dtest with tablets / compaction_additional_test.TestTimeWindowDataSegregation.test_streaming_on_rebuild_multidc (from dtest-with-tablets) to fail. E.g. as in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22771/testReport/junit/compaction_additional_test/TestTimeWindowDataSegregation/release___dtest_with_tablets___test_streaming_on_rebuild_multidc/
> 
> Correct way is for sure to add a dtest/skip-labels here, which can be done in advance even before merging the PR

Do we have skip label for master?

**Comment by pehala** [comment]
> Do we have skip label for master?

We do not, if it is not on master the issue should not be closed. Not sure what it would be useful for

**Comment by denesb** [comment]
> > Do we have skip label for master?
> 
> We do not, if it is not on master the issue should not be closed. Not sure what it would be useful for

I'm migrating issues to JIRA. Once I have the JIRA issue, I close the github one. This now makes the bad test run.

**Comment by pehala** [comment]
> I'm migrating issues to JIRA. Once I have the JIRA issue, I close the github one. This now makes the bad test run.

In that case, the update to the test will be needed :-( We should swap the github issues in the code for the JIRA ones

**Comment by denesb** [comment]
Closing, the dtests were fixed by https://github.com/scylladb/scylla-dtest/pull/6778.
